## Data preprocessing for ML Phase

In [59]:
import pandas as pd 
import numpy as np 

In [100]:
# To show all Columns
pd.set_option('display.max_columns', None)

In [101]:
## reading the cleaned data from Analysis phase but before missing handling to avoid data leakage 
df=pd.read_csv(r"cleaned_df_for_ML.csv",index_col=0)

In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 74111 entries, 0 to 74110
Data columns (total 63 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   log_price                                          74111 non-null  float64
 1   room_type                                          74111 non-null  object 
 2   accommodates                                       74111 non-null  int64  
 3   bathrooms                                          73911 non-null  float64
 4   bed_type                                           74111 non-null  object 
 5   cancellation_policy                                74111 non-null  object 
 6   cleaning_fee                                       74111 non-null  bool   
 7   city                                               74111 non-null  object 
 8   host_has_profile_pic                               73923 non-null  object 
 9   host_identi

In [102]:
df.head()

,log_price,room_type,accommodates,bathrooms,bed_type,cancellation_policy,cleaning_fee,city,host_has_profile_pic,host_identity_verified,host_response_rate,instant_bookable,latitude,longitude,neighbourhood,number_of_reviews,review_scores_rating,bedrooms,beds,amenities_count,has_internet,has_tv,has_air_conditioning,is_family_friendly,has_washer,has_dryer,has_kitchen,has_parking,has_elevator,has_heating,has_smoke_detector,has_carbon_monoxide_detector,has_fire_extinguisher,has_workspace,has_pets_allowed,has_pool,has_breakfast,has_first_aid_kit,amenity__wireless_internet,amenity__kitchen,amenity__heating,amenity__essentials,amenity__smoke_detector,amenity__air_conditioning,amenity__tv,amenity__shampoo,amenity__hangers,amenity__carbon_monoxide_detector,amenity__internet,amenity__laptop_friendly_workspace,amenity__hair_dryer,amenity__washer,amenity__dryer,amenity__iron,amenity__family_kid_friendly,amenity__fire_extinguisher,amenity__first_aid_kit,amenity__translation_missing_enhosting_amenity_50,amenity__cable_tv,amenity__free_parking_on_premises,amenity__translation_missing_enhosting_amenity_49,amenity__24_hour_check_in,amenity__lock_on_bedroom_door,host_experience_years
0,5.010635,Entire home/apt,3,1.0,Real Bed,strict,True,New York City,True,True,NaN,False,40.696524,-73.991617,Brooklyn Heights,2,100.0,1.0,1.0,9,True,False,True,True,False,True,True,False,False,True,False,False,False,False,False,False,False,False,True,True,True,True,False,True,False,False,False,False,False,False,True,False,False,True,True,False,False,True,False,False,False,False,False,13.659138
1,5.129899,Entire home/apt,7,1.0,Real Bed,strict,True,New York City,True,False,100.0,True,40.766115,-73.989040,Hell's Kitchen,6,93.0,3.0,3.0,15,True,False,True,True,True,True,True,False,False,True,True,False,True,False,False,False,False,False,True,True,True,True,True,True,False,True,True,False,False,False,True,True,True,True,True,True,False,True,False,False,False,False,False,8.427105
2,4.976734,Entire home/apt,5,1.0,Real Bed,moderate,True,New York City,True,True,100.0,True,40.808110,-73.943756,Harlem,10,92.0,1.0,3.0,19,True,True,True,True,False,True,True,False,False,True,True,True,True,True,False,False,True,False,True,True,True,True,True,True,True,True,True,True,False,True,True,False,False,True,True,True,False,True,True,False,False,False,False,9.075975
3,6.620073,Entire home/apt,4,1.0,Real Bed,flexible,True,San Francisco,True,True,NaN,False,37.772004,-122.431619,Lower Haight,0,NaN,2.0,2.0,15,True,True,True,False,True,True,True,False,False,True,True,True,True,False,False,False,False,True,True,True,True,True,True,False,True,False,False,True,True,False,False,True,True,False,False,True,True,False,True,False,False,False,False,10.595483
4,4.744932,Entire home/apt,2,1.0,Real Bed,moderate,True,Washington DC,True,True,100.0,True,38.925627,-77.034596,Columbia Heights,4,40.0,0.0,1.0,12,True,True,True,False,False,False,True,False,True,True,True,True,True,False,False,False,False,False,True,True,True,True,True,True,True,True,False,True,True,False,False,False,False,False,False,True,False,False,False,False,False,False,False,10.729637


In [ ]:
## i will drop the neighborhood as  during the cat encoding it will talk time 
df.drop(columns='neighbourhood',axis=1,inplace=True)

### 1) Split Data into Input & Output

In [104]:
x = df.drop('log_price', axis= 1) ## saving df without log_price 
y = df['log_price'] ## target Feature

### 2) Split Data into Train & Test

In [105]:
from sklearn.model_selection import train_test_split 

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= 0.2, random_state= 42)### test_size=0.2 → 20% of the data goes to testing

### 3) Handle Numerical

In [106]:
df.isna().mean().round(4).sort_values(ascending=False)*100

host_response_rate              24.69
review_scores_rating            22.56
bathrooms                        0.27
host_experience_years            0.25
host_has_profile_pic             0.25
                                ...  
has_smoke_detector               0.00
has_carbon_monoxide_detector     0.00
room_type                        0.00
has_workspace                    0.00
has_fire_extinguisher            0.00
Length: 63, dtype: float64

## Missing Values Handling Plan

Below is the recommended action for each column based on the percentage of missing values.

### 🟠 Columns to FILL (missing between 5% and 40%)
| Column              | Missing % | Action | Suggested Strategy |
|---------------------|-----------|--------|---------------------|
| host_response_rate   | 24.69%   | **Fill** | after deep diving i  decided  KNN |
| review_scores_rating | 22.56%   | **Fill** | after deep diving i  decided  KNN |

---

### 🟢 Columns to DROP ROWS (missing < 5%)
| Column                  | Missing % | Action |
|-------------------------|-----------|--------|
| bathrooms               | 0.27%     | **Drop rows** | 
| host_experience_years   | 0.25%     | **Drop rows** |
| host_identity_verified  | 0.25%     | **Drop rows** | 
| host_has_profile_pic    | 0.25%     | **Drop rows** |
| beds                    | 0.18%     | **Drop rows** | 
| bedrooms                | 0.12%     | **Drop rows** |


In [107]:
## drop  rows  for columns <5%

cols_to_drop_na = ['bathrooms', 'host_experience_years', 'host_identity_verified', 'host_has_profile_pic', 'beds', 'bedrooms']
x_train = x_train.dropna(subset=cols_to_drop_na).reset_index(drop=True)
x_test = x_test.dropna(subset=cols_to_drop_na).reset_index(drop=True)

In [109]:
x_train['host_has_profile_pic']=x_train['host_has_profile_pic'].astype(bool)
x_test['host_has_profile_pic']=x_test['host_has_profile_pic'].astype(bool)
x_train['host_identity_verified']=x_train['host_identity_verified'].astype(bool)
x_test['host_identity_verified']=x_test['host_identity_verified'].astype(bool)

In [113]:
x_train['host_has_profile_pic'].dtypes

dtype('bool')

#### 3.1) Impute Missing Values

In [114]:
num_cols = x_train.select_dtypes(include= 'number').columns
num_cols

Index(['accommodates', 'bathrooms', 'host_response_rate', 'latitude',
       'longitude', 'number_of_reviews', 'review_scores_rating', 'bedrooms',
       'beds', 'amenities_count', 'host_experience_years'],
      dtype='object')

In [115]:
x_train[num_cols].isna().mean().round(4).sort_values(ascending=False)*100

host_response_rate       24.48
review_scores_rating     22.58
accommodates              0.00
bathrooms                 0.00
latitude                  0.00
longitude                 0.00
number_of_reviews         0.00
bedrooms                  0.00
beds                      0.00
amenities_count           0.00
host_experience_years     0.00
dtype: float64

In [116]:
x_test[num_cols].isna().mean().round(4).sort_values(ascending=False)*100

host_response_rate       24.52
review_scores_rating     22.41
accommodates              0.00
bathrooms                 0.00
latitude                  0.00
longitude                 0.00
number_of_reviews         0.00
bedrooms                  0.00
beds                      0.00
amenities_count           0.00
host_experience_years     0.00
dtype: float64

In [117]:
# host_response_rate , review_scores_rating --> Fill with Knn Imputer
from sklearn.impute import KNNImputer

knn = KNNImputer()

x_train[num_cols] = knn.fit_transform(x_train[num_cols])
x_test[num_cols] = knn.transform(x_test[num_cols])

In [119]:
x_train.isna().mean().round(4).sort_values(ascending=False)*100## to check that there is no missing value any more 

room_type                0.0
amenity__internet        0.0
has_pool                 0.0
has_breakfast            0.0
has_first_aid_kit        0.0
                        ... 
has_parking              0.0
has_elevator             0.0
has_heating              0.0
has_smoke_detector       0.0
host_experience_years    0.0
Length: 62, dtype: float64

In [120]:
x_test.isna().mean().round(4).sort_values(ascending=False)*100## to check that there is no missing value any more 

room_type                0.0
amenity__internet        0.0
has_pool                 0.0
has_breakfast            0.0
has_first_aid_kit        0.0
                        ... 
has_parking              0.0
has_elevator             0.0
has_heating              0.0
has_smoke_detector       0.0
host_experience_years    0.0
Length: 62, dtype: float64

#### 3.2) Feature Scaling

In [122]:
# Use Robust Scaler as it is not affected by outliers
from sklearn.preprocessing import RobustScaler

robust_scaler = RobustScaler()

x_train[num_cols] = robust_scaler.fit_transform(x_train[num_cols])
x_test[num_cols] = robust_scaler.transform(x_test[num_cols])

In [123]:
x_train.select_dtypes(include= 'number')

,accommodates,bathrooms,host_response_rate,latitude,longitude,number_of_reviews,review_scores_rating,bedrooms,beds,amenities_count,host_experience_years
0,1.0,1.0,-0.810811,-0.990222,-0.929895,0.272727,0.125,1.0,2.0,0.000000,-0.150826
1,0.0,0.0,0.000000,0.022712,0.069349,0.000000,-1.125,0.0,0.0,0.111111,1.860537
2,1.0,0.0,0.000000,0.011342,0.068579,-0.090909,-2.000,-1.0,1.0,-0.666667,-1.034091
3,0.0,0.0,-3.054054,-0.438581,-1.023326,-0.272727,0.500,0.0,0.0,-1.888889,0.580579
4,0.5,0.0,0.000000,0.003140,0.068665,0.454545,0.375,0.0,0.0,-0.444444,1.198347
...,...,...,...,...,...,...,...,...,...,...,...
58840,0.0,1.0,-3.918919,-0.992363,-0.930315,-0.227273,0.500,0.0,0.0,-0.111111,0.878099
58841,-0.5,0.0,0.000000,-0.983956,-0.941477,-0.181818,-0.750,0.0,0.0,-0.111111,0.815083
58842,1.5,1.0,-0.675676,-0.993540,-0.931327,0.636364,-0.625,1.0,1.0,0.111111,0.875000
58843,0.0,0.0,0.000000,0.016469,0.068477,-0.272727,0.500,0.0,0.0,0.444444,1.352273


### 4) Handle Categorical

In [124]:
x_train.select_dtypes(include= 'object').head()

,room_type,bed_type,cancellation_policy,city
0,Entire home/apt,Real Bed,flexible,Los Angeles
1,Entire home/apt,Real Bed,strict,New York City
2,Entire home/apt,Real Bed,strict,New York City
3,Private room,Real Bed,flexible,San Francisco
4,Entire home/apt,Real Bed,moderate,New York City


In [125]:
cat_cols = x_train.select_dtypes(include= 'object').columns
cat_cols

Index(['room_type', 'bed_type', 'cancellation_policy', 'city'], dtype='object')

In [126]:
for col in cat_cols:

    print(col)
    print(df[col].nunique())
    print(df[col].unique())
    print('-' * 100)

room_type
3
['Entire home/apt' 'Private room' 'Shared room']
----------------------------------------------------------------------------------------------------
bed_type
5
['Real Bed' 'Futon' 'Pull-out Sofa' 'Couch' 'Airbed']
----------------------------------------------------------------------------------------------------
cancellation_policy
5
['strict' 'moderate' 'flexible' 'super_strict_30' 'super_strict_60']
----------------------------------------------------------------------------------------------------
city
6
['New York City' 'San Francisco' 'Washington DC' 'Los Angeles' 'Chicago'
 'Boston']
----------------------------------------------------------------------------------------------------


#### 4.1) Encoding Ordinal Columns

In [127]:
# cancellation_policy  encoding with ordinal encoder
from sklearn.preprocessing import OrdinalEncoder

ord_encoder = OrdinalEncoder(categories= [ ['super_strict_60', 'super_strict_30','strict', 'moderate', 'flexible'] ])

x_train[['cancellation_policy']] = ord_encoder.fit_transform(x_train[['cancellation_policy']])
x_test[['cancellation_policy']] = ord_encoder.transform(x_test[['cancellation_policy']])

In [128]:
x_train['cancellation_policy'].head()

0    4.0
1    2.0
2    2.0
3    4.0
4    3.0
Name: cancellation_policy, dtype: float64

#### 4.3) Encoding Nominal (OneHotEncoder)

In [129]:
nominal_cols = x_train.select_dtypes(include= 'object').columns
nominal_cols

Index(['room_type', 'bed_type', 'city'], dtype='object')

In [130]:
from sklearn.preprocessing import OneHotEncoder## i will use onehotencoder as the nominal feature less than 7 variables 

ohe = OneHotEncoder(drop= 'first', sparse_output= False)

ohe_arr_train = ohe.fit_transform(x_train[nominal_cols])
ohe_arr_test = ohe.transform(x_test[nominal_cols])

In [131]:
## creating dataframe from the result from onhotencoder 
ohe_df_train = pd.DataFrame(ohe_arr_train, columns = ohe.get_feature_names_out())
ohe_df_test = pd.DataFrame(ohe_arr_test, columns = ohe.get_feature_names_out())

In [132]:
# Reset index before concatenation
x_train.reset_index(drop= True, inplace= True)
x_test.reset_index(drop= True, inplace= True)

y_train.reset_index(drop= True, inplace= True)
y_test.reset_index(drop= True, inplace= True)

In [133]:
## concatenate the output to x_train , x_test then drop orginal columns 
x_train = pd.concat([x_train, ohe_df_train], axis= 1).drop(columns= nominal_cols, axis= 1)
x_test = pd.concat([x_test, ohe_df_test], axis= 1).drop(columns= nominal_cols, axis= 1)

In [134]:
x_train.head()

,accommodates,bathrooms,cancellation_policy,cleaning_fee,host_has_profile_pic,host_identity_verified,host_response_rate,instant_bookable,latitude,longitude,number_of_reviews,review_scores_rating,bedrooms,beds,amenities_count,has_internet,has_tv,has_air_conditioning,is_family_friendly,has_washer,has_dryer,has_kitchen,has_parking,has_elevator,has_heating,has_smoke_detector,has_carbon_monoxide_detector,has_fire_extinguisher,has_workspace,has_pets_allowed,has_pool,has_breakfast,has_first_aid_kit,amenity__wireless_internet,amenity__kitchen,amenity__heating,amenity__essentials,amenity__smoke_detector,amenity__air_conditioning,amenity__tv,amenity__shampoo,amenity__hangers,amenity__carbon_monoxide_detector,amenity__internet,amenity__laptop_friendly_workspace,amenity__hair_dryer,amenity__washer,amenity__dryer,amenity__iron,amenity__family_kid_friendly,amenity__fire_extinguisher,amenity__first_aid_kit,amenity__translation_missing_enhosting_amenity_50,amenity__cable_tv,amenity__free_parking_on_premises,amenity__translation_missing_enhosting_amenity_49,amenity__24_hour_check_in,amenity__lock_on_bedroom_door,host_experience_years,room_type_Private room,room_type_Shared room,bed_type_Couch,bed_type_Futon,bed_type_Pull-out Sofa,bed_type_Real Bed,city_Chicago,city_Los Angeles,city_New York City,city_San Francisco,city_Washington DC
0,1.0,1.0,4.0,False,True,True,-0.810811,False,-0.990222,-0.929895,0.272727,0.125,1.0,2.0,0.000000,True,True,True,False,True,True,True,False,False,True,True,False,True,False,False,True,False,False,True,True,True,True,True,True,True,False,False,False,True,False,False,True,True,False,False,True,False,True,True,False,True,False,False,-0.150826,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0
1,0.0,0.0,2.0,True,True,True,0.000000,True,0.022712,0.069349,0.000000,-1.125,0.0,0.0,0.111111,True,True,False,True,True,True,True,True,False,True,True,True,True,False,True,False,False,True,True,True,True,True,True,False,True,False,True,True,False,False,True,True,True,True,True,True,True,True,False,True,False,False,False,1.860537,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
2,1.0,0.0,2.0,True,True,False,0.000000,True,0.011342,0.068579,-0.090909,-2.000,-1.0,1.0,-0.666667,False,True,True,False,False,True,True,False,False,True,True,True,False,False,False,False,False,True,False,True,True,True,True,True,True,True,False,True,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,-1.034091,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
3,0.0,0.0,4.0,True,True,True,-3.054054,False,-0.438581,-1.023326,-0.272727,0.500,0.0,0.0,-1.888889,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,0.580579,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
4,0.5,0.0,3.0,True,True,True,0.000000,False,0.003140,0.068665,0.454545,0.375,0.0,0.0,-0.444444,True,False,True,False,False,False,True,True,False,True,False,False,False,False,False,False,False,False,True,True,True,True,False,True,False,True,False,False,True,False,False,False,False,False,False,False,False,True,False,True,True,False,False,1.198347,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0


### 5) Handling Imbalance